### ISSUE TITLE
transform() speed for online outlier detection

### ISSUE THREAD

**Issue by Timonzimm** [lmcinnes/umap#219]
Hi there!
I am having very good performances with UMAP on my dataset for outlier detection (Isolation forest after UMAP projection).
However, when I am trying to use only `transform()` on 1 sample after a `fit()` on 500 samples, it takes longer on the single sample setting than on the full data (500).

I saw another issue mentioning the fact that UMAP will just take as much time on the full dataset than one sample, but then it shouldn't be longer on 5000 samples, right? Well, the `transform()` on 5000 samples is in fact much slower so I do not understand how things work between fit and transform for UMAP.

Could anybody help me on this? I am trying to do some online outlier detection.

Best,
Timon

**Comment by lmcinnes** [comment]
Part of the catch here is that you are in the "small data" regime of UMAP, and the transform is (relatively) less efficient in this case. The transform method is really more designed for the case where you have, say, 100,000 samples, and wish to add 10 or 100 new samples. In practice adding 100 samples will be a lot faster than training the full 100,000 sample dataset. In the small data regime this is not so good, as there is non-trivial work required for the transform. If you have training data that small then UMAP may not necessarily be the best fit for your use case.

**Comment by monchin** [comment]
@lmcinnes Thank you for the awesome project.
I also found it seems that `transform()` is somehow much slower than `fit()`.
I have 2 numpy array, `data1` is about 400,000\*22, and `data2` is about 370,000*22. I first `fit(data1)`, and it cost 815s; then I `transform(data1)`, it just cost 0.1s; but when I `transform(data2)`, it cost about 1641s, although `data2` is smaller than `data1`. Then, I `np.random.shuffle(data1)`, and `transform(data1)`, this time it cost 2018s as `fit()`would just take 815s for a numpy array with the same shape. 
I maintain the model and use `transform()` because the layout of the two arrays are almost the same, so I'd like to get the same layout in 2-D space. Inference is much slower than train is a little strange.

**Comment by lmcinnes** [comment]
If you have a lot of data then fit is always the best option. There is a shortcut if you are transforming precisely the training data (hence the speed in that particular case), transform is simply to allow you to adapt small amounts of new data to an existing model. It is not classical inference -0- it is quite a costly process, but if you need to retain the existing model it is the only option available.

**Comment by nianlong93** [comment]
@lmcinnes may I ask how transform works conceptually? During tranform some new data, do we need to put the new data and the original high dimensional data together to find the K nearest neighbor, then perform gradient descent to only train the low dimension embeddings of the new data? If so, what I am confused is: when we do tranform, do we need keep the original high dimensional data in the background? Thanks!

**Comment by lmcinnes** [comment]
Yes, that is indeed the case -- we are storing an index for approximate nearest neighbor search, and that involves storing the original data.
